### PR TITLE
Http Feed module + Spamhaus feed

### PR DESCRIPTION
**Pull request by liorblob**
## Status
In Progress

## Related Issues
https://github.com/demisto/etc/issues/21019

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [Spamhaus Feed](https://github.com/demisto/content/pull/5394/files#diff-086e97388d2b9ba34702f8e951068421)

### CODE REVIEW

**Comment by liorblob** [comment]
@ronykoz @avidan-H @teizenman Please have a look and tell me if it looks right.

File: Integrations/SpamhausFeed/SpamhausFeed.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        if demisto.command() == 'fetch-indicators':
+            indicators = fetch_indicators_command(client, params.get('indicator_type'))
+            # we submit the indicators in batches
+            for b in batch(indicators, batch_size=2000):
+                demisto.createIndicators(b)
+        else:
+            readable_output, outputs, raw_response = commands[command](client, demisto.args())
+            return_outputs(readable_output, outputs, raw_response)
+    except Exception as e:
+        err_msg = f'Error in Spamhaus feed [{e}]'
```
Comment (ronykoz) [inline Integrations/SpamhausFeed/SpamhausFeed.py]:
```suggestion
        err_msg = f'Error in {FEED_NAME} feed [{e}]'
```

File: Integrations/SpamhausFeed/SpamhausFeed.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import demistomock as demisto
+from CommonServerPython import *
+from CommonServerUserPython import *
+
+
```
Comment (ronykoz) [inline Integrations/SpamhausFeed/SpamhausFeed.py]: FEED_NAME = 'Spamhaus'

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                fattrs['regex'] = re.compile(fattrs['regex'])
+            else:
+                raise ValueError(f'{self.source_name} - {f} field does not have a regex')
+            if 'transform' not in fattrs:
+                if fattrs['regex'].groups > 0:
+                    LOG(f'{self.source_name} - no transform string for field {f} but pattern contains groups')
+                fattrs['transform'] = r'\g<0>'
+
+    def build_iterator(self):
+        rkwargs = dict(
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: add an option to pass kwargs here

File: Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+FEED_NAME = 'Spamhaus'
+
+
+def main():
+    feed_main()
```
Comment (ronykoz) [inline Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.py]: add a comment explaining this

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    client.build_iterator()
+    return 'ok', {}, {}
+
+
+def feed_main():
+    # Write configure here
+    params = {k: v for k, v in demisto.params().items() if v is not None}
+    client = Client(**params)
+    command = demisto.command()
+    demisto.info('Command being called is {}'.format(command))
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: Isn't it problematic for the fetch-indicators scenario?

Comment (liorblob) [reply Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: Problematic how?

**Comment by kirbles19** [comment]
@liorblob Done.

**Comment by liorblob** [comment]
@ronykoz can you approve?

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            if 'regex' in fattrs:
+                fattrs['regex'] = re.compile(fattrs['regex'])
+            else:
+                raise ValueError(f'{self.feed_name} - {f} field does not have a regex')
+            if 'transform' not in fattrs:
+                if fattrs['regex'].groups > 0:
+                    LOG(f'{self.feed_name} - no transform string for field {f} but pattern contains groups')
+                fattrs['transform'] = r'\g<0>'
+
+    def build_iterator(self):
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: add docstring and explain the meaning of this, not sure that the name is good enough

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    try:
+        if command == 'fetch-indicators':
+            indicators = fetch_indicators_command(client, params.get('indicator_type'))
+            # we submit the indicators in batches
+            for b in batch(indicators, batch_size=2000):
+                demisto.createIndicators(b)
+        else:
+            args = demisto.args()
+            args['feed_name'] = feed_name
+            args['default_indicator_type'] = params.get('indicator_type')
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: those params are in the client already, why adding them to args?

File: Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.yml [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+- defaultvalue: '240'
+  display: Fetch Interval
+  name: feedFetchInterval
+  required: true
+  type: 19
+- additionalinfo: Type of the indicator in the feed.
+  display: Indicator Type
+  name: indicator_type
+  required: true
+  type: 0
```
Comment (avidan-H) [inline Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.yml]: It's kindof basa that we need to include this basically for only in the case of the generic HTTP Feed integration... whereas every specific implementation that uses the http module should in-theory have the indicator type hardcoded as part of the feed_type dictionary...

File: Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.yml [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+- name: expiration
+  required: false
+  type: 17
+- additionalinfo: When selected, the exclusion list will be ignored for this feed.
+  display: Bypass exclusion list
+  name: skipExclusionList
+  required: false
+  type: 8
+- defaultvalue: feedInstanceReputationNotSet
+  display: Mark as reputation
```
Comment (avidan-H) [inline Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.yml]:
```suggestion
  display: Mark with reputation
```

Comment (avidan-H) [reply Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.yml]: nevermind this is something that server should update for BYOI integrations that fetch indicators.

File: Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.yml [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+commonfields:
+  id: SpamhausFeed
+  version: -1
+configuration:
+- display: Fetch indicators
+  name: feed
+  required: false
+  type: 8
+- defaultvalue: '240'
+  display: Fetch Interval
```
[the remark this answers is not in the document]
Comment (avidan-H) [reply Packs/SpamhausFeed/Integrations/SpamhausFeed/SpamhausFeed.yml]: shoot. this is also something that should be changed on server side I think.

**Comment by avidan-H** [comment]
Just remember that the generic HTTP Feed integration needs to be updated to use the module after this gets merged.

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+                urls = [urls]
+            for url in urls:
+                r = requests.get(
+                    url,
+                    **kwargs
+                )
+                try:
+                    r.raise_for_status()
+                except Exception:
+                    LOG(f'{self.feed_name!r} - exception in request:'
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: self.feed_name!r ?

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+        self.indicator_type = indicator_type
+        if feed_url_to_config:
+            self.feed_url_to_config = feed_url_to_config
+        else:
+            self.feed_url_to_config = {url: self.get_feed_config(fields, indicator)}
+        self.ignore_regex: Optional[Pattern] = None
+        if ignore_regex is not None:
+            self.ignore_regex = re.compile(ignore_regex)
+
+    def get_feed_config(self, fields_json: str = '', indicator_json: str = ''):
```
Comment (ronykoz) [inline Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: whats the diff between indicator_json and fields_json add examples

File: Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py (hunk 11 above)
Comment (liorblob) [reply Packs/ApiModules/Scripts/HTTPFeedApiModule/HTTPFeedApiModule.py]: it's `repr` for mypy str-bytes error

**Pull request merged**